Plumbing engine dev

wendi-yu opened the pull request:
Created constructors for PlumbingEngine and PlumbingComponent classes, as well as (public) helper functions that are called when initializing. Basic functionality works but edge cases haven't been tested yet.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5)
<!-- Reviewable:end -->

Commit:
Set up constructors for each class

Constructors for PlumbingEngine and PlumbingComponent classes, as well
as function to load in graphs. Basic functionality works but haven't
considered edge cases, invalid input, etc. yet.

Commit:
Move functionality to helper functions

_set_time_resolution() and toggle_component() now handle some
reusable constructor functionality

Commit:
Apply formatting to code

jacobdeery reviewed (commented):
I didn't look at the unit tests yet because I know you're still working on those, but here are my initial review comments. Good job overall, especially with keeping the networkx interactions from getting ridiculously cryptic.

> Reviewed 2 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-:-M4I5uO_2TfCMXuk7Xm3:bkxh5sc)* status: 2 of 3 files reviewed, 11 unresolved discussions (waiting on @jacobdeery and @wendi-yu)

___
*[operations_simulator/plumbing/config.py, line 1 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I-R7k59dwomWWF5p3:-M4I-R7k59dwomWWF5p4:b-or5g5q) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/config.py#L1)):*
> ```Python
> '''config settings'''
> ```

Capital `C`

This docstring also doesn't add a ton right now, since it's already clear from the name that this holds config settings. At the same time, there isn't much else you can add to the docstring. See a comment further down about making this into a `utils` file instead that also holds util plumbing functions.
___
*[operations_simulator/plumbing/config.py, line 2 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I4NC46vwD6CBbYwpl:-M4I4NC46vwD6CBbYwpm:b-r8e3st) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/config.py#L2)):*
> ```
> DEFAULT_TIME_RESOLUTION = 0.5
> ```
perhaps `DEFAULT_TIME_RESOLUTION_S` to make it clear what the units are
___
*[operations_simulator/plumbing/plumbing_engine.py, line 6 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4HzxRo0PrKwuChsKAu:-M4HzxRo0PrKwuChsKAv:b-kjyiol) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L6)):*
> ```
> def teq_to_FC(teq):
> ```
I would suggest moving these functions (along with potentially the stuff currently in `config.py` into a standalone `plumbing_utils.py` file.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 17 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4HzlpaCAVAtEK_Dkxi:-M4HzlpbAPU9eRgAQTw6:b-w3hdlj) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L17)):*
> ```
> def FC_to_teq(FC):
>     return 4.5 / FC
> ```

Need special logic for this or you can divide by 0
___
*[operations_simulator/plumbing/plumbing_engine.py, line 60 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I0Cj351YfS9GUXVOg:-M4I0Cj351YfS9GUXVOh:b-8sytb7) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L60)):*
> ```Python
>             self.toggle_component(component, state_id)
> 
>     def toggle_component(self, component, state_id):
> ```

`set_component_state` might be a more clear name now that we're not just automatically toggling to the next state
___
*[operations_simulator/plumbing/plumbing_engine.py, line 61 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I-Oq64UZXAWcuwL4U:-M4I-Oq64UZXAWcuwL4V:bq1mqcg) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L61)):*
> ```Python
> 
>     def toggle_component(self, component, state_id):
>         '''change a component's state on the main graph'''
> ```

Capital `C`
___
*[operations_simulator/plumbing/plumbing_engine.py, line 67 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I45zb-p9nH5Sduxho:-M4I45zb-p9nH5Sduxhp:b-jv9x2o) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L67)):*
<details><summary><i alt="summary">Quoted 14 lines of code…</i></summary>

> ```
>         # Dict of {edges:FC} with component node names
>         state_edges_component = component.states[state_id]
> 
>         # Create new dict keyed by graph edges rather than component ones
>         state_edges_graph = {}
>         for cedge in state_edges_component.keys():
>             cstart_node, cend_node, key = cedge
>             new_edge = (component_map[cstart_node],
>                         component_map[cend_node], key)
>             state_edges_graph[new_edge] = state_edges_component[cedge]
> 
>         # set FC on main graph according to new dict
>         nx.classes.function.set_edge_attributes(
>             self.plumbing_graph, state_edges_graph, 'FC')
> ```
</details>

This ended up being a lot more complicated than I thought it would be. Nice job keeping the implementation relatively intuitive.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 83 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I0KKZ5fqK35LWi2Me:-M4I0KKZ5fqK35LWi2Mf:bq1mqq8) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L83)):*
> ```Python
> 
>     def _set_time_resolution(self, component):
>         '''set a time resolution based on lowest teq (highest FC)'''
> ```

Capital `S`
___
*[operations_simulator/plumbing/plumbing_engine.py, line 93 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I6Dbn93AgbiYOdoeC:-M4I6Dbn93AgbiYOdoeD:b-2bov4w) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L93)):*
> ```
>             self.time_resolution = FC_to_teq(
>                 max_fc) / config.DEFAULT_RESOLUTION_SCALE
> ```
hmm, I'm not sure why autopep8 wrapped this line for you. It _should_ be configured to allow line lengths up to 100 characters. I'll look into it.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 101 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I2thn4knGSC1Mudqc:-M4I2thn4knGSC1Mudqd:bzew86s) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L101)):*
> ```Python
>         self.component_graph = nx.MultiDiGraph(edge_list)
>         self.states = states
>         self.current_state = ''
> ```

Should we use `None` instead of the empty string to indicate no state?
___
*[operations_simulator/plumbing/plumbing_engine.py, line 106 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I33Vm47L-lUt4OUox:-M4I33Vm47L-lUt4OUoy:b9tzyy7) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L106)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
>                 if state[edge] == 0:
>                     state[edge] = config.FC_MAX
>                 elif state[edge] == config.CLOSED_KEYWORD:
>                     state[edge] = 0
>                 else:
>                     state[edge] = teq_to_FC(state[edge])
> ```
</details>

This logic should all be in `teq_to_FC`


<!-- Sent from Reviewable.io -->

Commit:
Change time_resolution to operate in micro_s, address PR comments

wendi-yu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5)* status: 0 of 3 files reviewed, 11 unresolved discussions (waiting on @jacobdeery)

___
*[operations_simulator/plumbing/config.py, line 1 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I-R7k59dwomWWF5p3:-M4L4U766q5485PVLOaS:becu3zm) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/config.py#L1)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Capital `C`

This docstring also doesn't add a ton right now, since it's already clear from the name that this holds config settings. At the same time, there isn't much else you can add to the docstring. See a comment further down about making this into a `utils` file instead that also holds util plumbing functions.
</blockquote></details>

Done, though the capital in an unpushed commit right now 
___
*[operations_simulator/plumbing/config.py, line 2 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I4NC46vwD6CBbYwpl:-M4L3pvg-NDAbqqIRajS:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/config.py#L2)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> DEFAULT_TIME_RESOLUTION = 0.5
> ```
perhaps `DEFAULT_TIME_RESOLUTION_S` to make it clear what the units are
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 6 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4HzxRo0PrKwuChsKAu:-M4L3uJH02ifQJhyUnED:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L6)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> def teq_to_FC(teq):
> ```
I would suggest moving these functions (along with potentially the stuff currently in `config.py` into a standalone `plumbing_utils.py` file.
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 17 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4HzlpaCAVAtEK_Dkxi:-M4L3taE0tKGznDAEiFp:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L17)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> def FC_to_teq(FC):
>     return 4.5 / FC
> ```

Need special logic for this or you can divide by 0
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 60 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I0Cj351YfS9GUXVOg:-M4L3x4V3Sv73kw-20mc:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L60)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

`set_component_state` might be a more clear name now that we're not just automatically toggling to the next state
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 61 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I-Oq64UZXAWcuwL4U:-M4L44hEAMQc0Ggcn8Se:bsez4cu) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L61)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Capital `C`
</blockquote></details>

Done, though it's in an unpushed commit right now.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 67 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I45zb-p9nH5Sduxho:-M4L3ybI3QWQWsizLzc-:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L67)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
>         # Dict of {edges:FC} with component node names
>         state_edges_component = component.states[state_id]
> 
>         # Create new dict keyed by graph edges rather than component ones
>         state_edges_graph = {}
>         for cedge in state_edges_component.keys():
>             cstart_node, cend_node, key = cedge
>             new_edge = (component_map[cstart_node],
>                         component_map[cend_node], key)
>             state_edges_graph[new_edge] = state_edges_component[cedge]
> 
>         # set FC on main graph according to new dict
>         nx.classes.function.set_edge_attributes(
>             self.plumbing_graph, state_edges_graph, 'FC')
> ```
This ended up being a lot more complicated than I thought it would be. Nice job keeping the implementation relatively intuitive.
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 83 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I0KKZ5fqK35LWi2Me:-M4L4dWNCfI6i8LtEbjI:bsiw3cs) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L83)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Capital `S`
</blockquote></details>

Done, though it's in an unpushed commit right now.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 93 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I6Dbn93AgbiYOdoeC:-M4L4-IFCEhNN9TPUXdB:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L93)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
>             self.time_resolution = FC_to_teq(
>                 max_fc) / config.DEFAULT_RESOLUTION_SCALE
> ```
hmm, I'm not sure why autopep8 wrapped this line for you. It _should_ be configured to allow line lengths up to 100 characters. I'll look into it.
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 101 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I2thn4knGSC1Mudqc:-M4L409f1Es8HLpaNnXa:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L101)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Should we use `None` instead of the empty string to indicate no state?
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 106 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I33Vm47L-lUt4OUox:-M4L40vB8dNwsBWQLh9X:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L106)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
>                 if state[edge] == 0:
>                     state[edge] = config.FC_MAX
>                 elif state[edge] == config.CLOSED_KEYWORD:
>                     state[edge] = 0
>                 else:
>                     state[edge] = teq_to_FC(state[edge])
> ```
This logic should all be in `teq_to_FC`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jacobdeery reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5)* status: 0 of 3 files reviewed, 12 unresolved discussions (waiting on @jacobdeery and @wendi-yu)

___
*[operations_simulator/plumbing/plumbing_engine.py, line 3 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L9jutEzKwkRkJlbtl:-M4L9jutEzKwkRkJlbtm:b-unkr4h) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_engine.py#L3)):*
> ```
> import networkx as nx
> import matplotlib.pyplot as plt
> import plumbing_utils as utils
> ```

Two things:
1. Third-party imports should be separated from local imports by a newline
2. `import plumbing_utils as utils` is an implicit relative import that will error unless you are running the script from this directory. This should either be `import .plumbing_utils as utils` or `import operations_simulator.plumbing.plumbing_utils as utils`.

PEP 8 on imports: https://www.python.org/dev/peps/pep-0008/#imports
___
*[operations_simulator/plumbing/plumbing_engine.py, line 61 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L7CXI5SVA5KqKauSq:-M4L7CXI5SVA5KqKauSr:b-qfotdn) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_engine.py#L61)):*
> ```
>             new_edge = (component_map[cstart_node],
>                         component_map[cend_node], key)
> ```
Nit: this should now be able to fit on one line
___
*[operations_simulator/plumbing/plumbing_engine.py, line 66 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L7F4z3tjkzWa4W6Oq:-M4L7F4z3tjkzWa4W6Or:b-27vjeq) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_engine.py#L66)):*
> ```
>         nx.classes.function.set_edge_attributes(
>             self.plumbing_graph, state_edges_graph, 'FC')
> ```
Nit: here too
___
*[operations_simulator/plumbing/plumbing_engine.py, line 70 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L6pPG9Z7QQW_FOZe7:-M4L6pPG9Z7QQW_FOZe8:b7jk3zj) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_engine.py#L70)):*
> ```
>         max_fc = utils.teq_to_FC(self.time_resolution *
>                                  utils.DEFAULT_RESOLUTION_SCALE)
> ```
Nit: here too
___
*[operations_simulator/plumbing/plumbing_engine.py, line 91 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4KMDFC6xw8w-0480Ru:-M4KMDFC6xw8w-0480Rv:b-9mpsjn) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_engine.py#L91)):*
> ```Python
>         for state in self.states.values():
>             for edge in state:
>                 # TODO(jacob or wendi) figure out how exactly this error will be thrown
> ```

Nit: usually I would format the TODO like this:

`# TODO(jacob/wendi): Figure out how exactly this error will be thrown.`

I realize that the example I gave you on Slack doesn't follow this format, so that's my bad.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 95 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L6Z28Dj7lj8fdjVXT:-M4L6Z28Dj7lj8fdjVXU:b-2qd9sa) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_engine.py#L95)):*
> ```Python
>                     print("Warning: provided teq value too low")
>                     state[edge] = utils.MIN_TEQ
>                 state[edge] = utils.teq_to_FC(state[edge] * 1e6)
> ```

This can probably be a TODO, but I wonder if we want to use the `datetime.timedelta` class to make the time manipulation easier. No action required immediately, but something to keep in mind.

See docs: https://docs.python.org/3/library/datetime.html#timedelta-objects
___
*[operations_simulator/plumbing/plumbing_utils.py, line 1 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L7mvqE8DhvymmGooJ:-M4L7mvqE8DhvymmGooK:beg0idv) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_utils.py#L1)):*
> ```Python
> DEFAULT_TIME_RESOLUTION_MICROS = 10000
> ```

If we want to make this file available as part of the library, we need to add a line to `plumbing/__init__.py` to expose it, like this:

`from operations_simulator.plumbing.plumbing_utils import *`
___
*[operations_simulator/plumbing/plumbing_utils.py, line 20 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L5noA54V-wxs7Fgxv:-M4L5noA54V-wxs7Fgxw:b542tjp) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_utils.py#L20)):*
> ```
>         return 'closed'
> ```
`return CLOSED_KEYWORD`


<!-- Sent from Reviewable.io -->

Commit:
Add basic unit tests, address PR comments

Fixed some formatting/importing issues from PR comments. Also added unit
tests that test basic functionality; error tests will follow in a later
commit after we implement error catching in the functions.

wendi-yu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5)* status: 0 of 5 files reviewed, 8 unresolved discussions (waiting on @jacobdeery)

___
*[operations_simulator/plumbing/plumbing_engine.py, line 61 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I-Oq64UZXAWcuwL4U:-M4MSLZzECkXL088GCtu:b-q59rq4) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L61)):*
<details><summary><i>Previously, wendi-yu wrote…</i></summary><blockquote>

Done, though it's in an unpushed commit right now.
</blockquote></details>

Done now!
___
*[operations_simulator/plumbing/plumbing_engine.py, line 83 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4I0KKZ5fqK35LWi2Me:-M4MS_76DwO_uHEAx7A3:b-q59rq4) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/09863f7eeba5eea6a375fb5c164a81a265ef41cd/operations_simulator/plumbing/plumbing_engine.py#L83)):*
<details><summary><i>Previously, wendi-yu wrote…</i></summary><blockquote>

Done, though it's in an unpushed commit right now.
</blockquote></details>

Done now!
___
*[operations_simulator/plumbing/plumbing_engine.py, line 3 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L9jutEzKwkRkJlbtl:-M4MSK2B5voaD60YhHK7:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_engine.py#L3)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> import networkx as nx
> import matplotlib.pyplot as plt
> import plumbing_utils as utils
> ```

Two things:
1. Third-party imports should be separated from local imports by a newline
2. `import plumbing_utils as utils` is an implicit relative import that will error unless you are running the script from this directory. This should either be `import .plumbing_utils as utils` or `import operations_simulator.plumbing.plumbing_utils as utils`.

PEP 8 on imports: https://www.python.org/dev/peps/pep-0008/#imports
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 95 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L6Z28Dj7lj8fdjVXT:-M4MSeYXDU0ozJC5rWnw:bshmkue) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_engine.py#L95)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

This can probably be a TODO, but I wonder if we want to use the `datetime.timedelta` class to make the time manipulation easier. No action required immediately, but something to keep in mind.

See docs: https://docs.python.org/3/library/datetime.html#timedelta-objects
</blockquote></details>

Added a TODO in the comments.
___
*[operations_simulator/plumbing/plumbing_utils.py, line 1 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L7mvqE8DhvymmGooJ:-M4MSha85bhj6tEo_JMY:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_utils.py#L1)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

If we want to make this file available as part of the library, we need to add a line to `plumbing/__init__.py` to expose it, like this:

`from operations_simulator.plumbing.plumbing_utils import *`
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_utils.py, line 20 at r2](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4L5noA54V-wxs7Fgxv:-M4MSkChCR9wjy3JF8I9:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/569468e5f81c25f2a9538a8732d9ff1b121d8d9a/operations_simulator/plumbing/plumbing_utils.py#L20)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
>         return 'closed'
> ```
`return CLOSED_KEYWORD`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jacobdeery reviewed (commented):
I think we're almost there! Just one more thing in addition to my comments in the review: it would be good to have tests for `plumbing_utils` as well.

> Reviewed 1 of 3 files at r2, 2 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-:-M4M_Wmn92nT32dbhoOi:bx081kr)* status: 2 of 5 files reviewed, 10 unresolved discussions (waiting on @wendi-yu)

___
*[operations_simulator/plumbing/plumbing_engine.py, line 93 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MTXRRAZNGWpnH_KwE:-M4MTXRS5ZL4TX57EMYe:b-14293j) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/plumbing_engine.py#L93)):*
> ```
>                 # TODO(jacob/wendi): Look into eventually implementing this with datetime.timedelta object.
> ```

Nit: This line is longer than 100 characters, but unfortunately autopep8 won't automatically wrap it for us because it's an uninterrupted comment (https://github.com/hhatto/autopep8/issues/39). In cases like this we need to wrap it manually.
___
*[operations_simulator/plumbing/plumbing_utils.py, line 30 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MU0twCaEIwUMK3K6a:-M4MU0twCaEIwUMK3K6b:b56hlcu) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/plumbing_utils.py#L30)):*
> ```
> def s_to_micros_int(
>         sec):
> ```
Nit: this should be one line

Why the `int` in the function name?
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 3 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MVQHT-cK01FaeR9rx:-M4MVQHT-cK01FaeR9ry:b-wsu83j) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L3)):*
> ```Python
> import operations_simulator as ops
> import operations_simulator.plumbing.plumbing_utils as utils
> ```

Nit: missing an extra blank line here
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 4 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MUlteCYMGzF4bCXTC:-M4MUlteCYMGzF4bCXTD:b-mr6c8u) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L4)):*
<details><summary><i alt="summary">Quoted 13 lines of code…</i></summary>

> ```
> def two_edge_states_edges(s1e1, s1e2, s2e1, s2e2):
>     states = {
>         'open': {
>             (1, 2, 'A1'): s1e1,
>             (2, 1, 'A2'): s1e2
>         },
>         'closed': {
>             (1, 2, 'A1'): s2e1,
>             (2, 1, 'A2'): s2e2
>         }
>     }
>     edges = [(1, 2, 'A1'), (2, 1, 'A2')]
>     return (states, edges)
> ```
</details>


Writing helper functions like this is _awesome_ and a great way to make your tests less verbose. Nice job.
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 17 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MVN0LBvqUheBQ2kwn:-M4MVN0LBvqUheBQ2kwo:b-wsu83j) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L17)):*
> ```Python
>     edges = [(1, 2, 'A1'), (2, 1, 'A2')]
>     return (states, edges)
> ```

Nit: missing an extra blank line here
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 20 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MVksa3mHE6EwecAxo:-M4MVksa3mHE6EwecAxp:b-frd9ro) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L20)):*
> ```
> def test_plumbing_component_exists():
>     plumb_comp = ops.PlumbingComponent('', {}, [])
>     assert plumb_comp is not None
> ```

I know I wrote the equivalent of this for `PlumbingEngine`, but that was just because the class didn't have any functionality to write an actual meaningful test for. On its own this test is a bit incomplete - I would turn it into something like your `test_empty_graph` from the `PlumbingEngine` tests.
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 41 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MWQv604BToANgztoC:-M4MWQv604BToANgztoD:boyv4mw) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L41)):*
<details><summary><i alt="summary">Quoted 9 lines of code…</i></summary>

> ```
>     assert pc1.states == {
>         'open': {
>             (1, 2, 'A1'): utils.FC_MAX,
>             (2, 1, 'A2'): utils.FC_MAX
>         },
>         'closed': {
>             (1, 2, 'A1'): 0,
>             (2, 1, 'A2'): 0}
>         }
> ```
</details>


Nit: the closing braces on `closed` and on the outer dict are formatted a bit weird.
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 42 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MWZMa3HIcaUz_6evC:-M4MWZMa3HIcaUz_6evD:bqmobrk) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L42)):*
> ```Python
>             (2, 1, 'A2'): 0}
>         }
> ```

Nit: missing extra newline
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 44 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MXdFaBnGP2ThXG-Xc:-M4MXdFaBnGP2ThXG-Xd:bq8zj03) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L44)):*
> ```
> 0.1, 12, 57, 0.00000000000001
> ```

Right now it's not immediately obvious why these particular values were chosen. I'm assuming that `0.00000000000001` is the special one here? If so, I would make it into a variable and give it a name so that it's a bit more self-documenting. I think something like this would be a bit more clear:

```
normal_teq = 1
teq_too_low = utils.millis_to_s(TEQ_MIN) / 2
states, edges = two_edge_states_edges(normal_teq, normal_teq, normal_teq, teq_too_low)
```
___
*[operations_simulator/plumbing/tests/test_plumbing_engine.py, line 53 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MZLL71OtTMOT5sp50:-M4MZLL71OtTMOT5sp51:bnego82) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_engine.py#L53)):*
> ```
> def test_plumbing_engine_exists():
>     plumb = ops.PlumbingEngine()
>     assert plumb is not None
> ```

You can delete this test now that you've written a better one with `test_empty_graph`.


<!-- Sent from Reviewable.io -->

Commit:
Add tests for plumbing utils, clean up code

wendi-yu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5)* status: 1 of 6 files reviewed, 4 unresolved discussions (waiting on @jacobdeery)

___
*[operations_simulator/plumbing/plumbing_utils.py, line 30 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MU0twCaEIwUMK3K6a:-M4Mph0O1Ze10TTfimBf:b-q0jz75) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/plumbing_utils.py#L30)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> def s_to_micros_int(
>         sec):
> ```
Nit: this should be one line

Why the `int` in the function name?
</blockquote></details>

Holdover from a previous implementation I forgot to remove. Done now.
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 4 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MUlteCYMGzF4bCXTC:-M4MqQh93vsRgD5Ivs1X:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L4)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> def two_edge_states_edges(s1e1, s1e2, s2e1, s2e2):
>     states = {
>         'open': {
>             (1, 2, 'A1'): s1e1,
>             (2, 1, 'A2'): s1e2
>         },
>         'closed': {
>             (1, 2, 'A1'): s2e1,
>             (2, 1, 'A2'): s2e2
>         }
>     }
>     edges = [(1, 2, 'A1'), (2, 1, 'A2')]
>     return (states, edges)
> ```

Writing helper functions like this is _awesome_ and a great way to make your tests less verbose. Nice job.
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 20 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MVksa3mHE6EwecAxo:-M4Mq-q04K3r6Z_DL9Da:b-g7ei6e) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L20)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> def test_plumbing_component_exists():
>     plumb_comp = ops.PlumbingComponent('', {}, [])
>     assert plumb_comp is not None
> ```

I know I wrote the equivalent of this for `PlumbingEngine`, but that was just because the class didn't have any functionality to write an actual meaningful test for. On its own this test is a bit incomplete - I would turn it into something like your `test_empty_graph` from the `PlumbingEngine` tests.
</blockquote></details>

Done - we won't support empty components as discussed.
___
*[operations_simulator/plumbing/tests/test_plumbing_component.py, line 44 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MXdFaBnGP2ThXG-Xc:-M4MqOQ80fiVQU_cd61U:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_component.py#L44)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> 0.1, 12, 57, 0.00000000000001
> ```

Right now it's not immediately obvious why these particular values were chosen. I'm assuming that `0.00000000000001` is the special one here? If so, I would make it into a variable and give it a name so that it's a bit more self-documenting. I think something like this would be a bit more clear:

```
normal_teq = 1
teq_too_low = utils.millis_to_s(TEQ_MIN) / 2
states, edges = two_edge_states_edges(normal_teq, normal_teq, normal_teq, teq_too_low)
```
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/tests/test_plumbing_engine.py, line 53 at r3](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MZLL71OtTMOT5sp50:-M4MqPNDAEuxndoQC8eR:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/47e4efb587b462f05c7d18c8f5318e344e1ae3c0/operations_simulator/plumbing/tests/test_plumbing_engine.py#L53)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> def test_plumbing_engine_exists():
>     plumb = ops.PlumbingEngine()
>     assert plumb is not None
> ```

You can delete this test now that you've written a better one with `test_empty_graph`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jacobdeery reviewed (commented):
Just a few nits left (and one problem that _does_ need to be fixed) and then we're good to merge. Please make sure that the application still works when packaged (although nothing in this PR should have broken that).

> Reviewed 4 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-:-M4Mu8JcBE5aikGa5vla:b-spfa1x)* status: 5 of 6 files reviewed, 5 unresolved discussions (waiting on @jacobdeery and @wendi-yu)

___
*[operations_simulator/plumbing/plumbing_engine.py, line 81 at r4](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4Ms0c64dmrlGKphHUB:-M4Ms0c64dmrlGKphHUC:bwuc6wg) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/b82848071aa1db3df2d48ae2d5079d4985f6a131/operations_simulator/plumbing/plumbing_engine.py#L81)):*
> ```
> eg
> ```

Super-nit: `e.g.` (although "such as" might be even better)

There's also a space at the beginning of the docstring.

I also don't know if "a component of a subsystem" is the best way to describe it - it's really "a discrete plumbing component", since the class doesn't hold any information about the system it's a part of and it may in fact not be part of any subsystem at all.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 101 at r4](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MsyqqBDOvbbkWWxKa:-M4MsyqqBDOvbbkWWxKb:b-eq2263) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/b82848071aa1db3df2d48ae2d5079d4985f6a131/operations_simulator/plumbing/plumbing_engine.py#L101)):*
> ```
> "Warning: provided teq value too low, replaced by minimum value: {}s"
> ```

Nit that I missed last time: Use single quotes here to stay consistent with everything else
___
*[operations_simulator/plumbing/tests/test_plumbing_utils.py, line 10 at r4](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MuNjY8UFhpmKaxOKJ:-M4MuNjY8UFhpmKaxOKK:babd2pi) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/b82848071aa1db3df2d48ae2d5079d4985f6a131/operations_simulator/plumbing/tests/test_plumbing_utils.py#L10)):*
> ```
>  
> ```

Nit: extra space.
___
*[operations_simulator/plumbing/tests/test_plumbing_utils.py, line 33 at r4](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4Mqgz42uDbmZL7xjAp:-M4Mqgz42uDbmZL7xjAq:b3ispqo) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/b82848071aa1db3df2d48ae2d5079d4985f6a131/operations_simulator/plumbing/tests/test_plumbing_utils.py#L33)):*
> ```
> assert utils.micros_to_s(i) - (i / 1e6) < epsilon
> ```

Needs an `abs` (`assert abs(utils.micros_to_s(i) - (i / 1e6)) < epsilon`), otherwise this lets large negative difference slips through.

You can also use `math.isclose` if you prefer: https://docs.python.org/3/library/math.html#math.isclose
___
*[operations_simulator/plumbing/tests/test_plumbing_utils.py, line 42 at r4](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MrHSA-tNLUC0s1NyA:-M4MrHSB1J0qVWgZsMAW:b-ya81ng) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/b82848071aa1db3df2d48ae2d5079d4985f6a131/operations_simulator/plumbing/tests/test_plumbing_utils.py#L42)):*
> ```Python
>     i = 0
>     while i < large_second_amount:
>         assert utils.s_to_micros(i) - (i * 1e6) < epsilon
> ```

Same as above


<!-- Sent from Reviewable.io -->

wendi-yu reviewed (commented):
Yep, both main and the executable still work after deleting `build` and rerunning `build.sh`. 

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-:-M4NEkbE3At2Evh5rY9Y:bqqcj0o)* status: 5 of 6 files reviewed, 2 unresolved discussions (waiting on @jacobdeery)

___
*[operations_simulator/plumbing/tests/test_plumbing_utils.py, line 33 at r4](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4Mqgz42uDbmZL7xjAp:-M4MyUGO9sJcX8_6lYvL:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/b82848071aa1db3df2d48ae2d5079d4985f6a131/operations_simulator/plumbing/tests/test_plumbing_utils.py#L33)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> assert utils.micros_to_s(i) - (i / 1e6) < epsilon
> ```

Needs an `abs` (`assert abs(utils.micros_to_s(i) - (i / 1e6)) < epsilon`), otherwise this lets large negative difference slips through.

You can also use `math.isclose` if you prefer: https://docs.python.org/3/library/math.html#math.isclose
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/tests/test_plumbing_utils.py, line 42 at r4](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MrHSA-tNLUC0s1NyA:-M4MyUq5DEUNPKTuPycZ:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/b82848071aa1db3df2d48ae2d5079d4985f6a131/operations_simulator/plumbing/tests/test_plumbing_utils.py#L42)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Same as above
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
Clean up code and fix float equality bug

jacobdeery reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5)* status: 4 of 6 files reviewed, all discussions resolved (waiting on @jacobdeery)

___
*[operations_simulator/plumbing/tests/test_plumbing_utils.py, line 10 at r4](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-M4MuNjY8UFhpmKaxOKJ:-M4PTI-_2Mfj1DVvdcln:b-f7oj2d) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/b82848071aa1db3df2d48ae2d5079d4985f6a131/operations_simulator/plumbing/tests/test_plumbing_utils.py#L10)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
>  
> ```

Nit: extra space.
</blockquote></details>

Not the newline; the space in `assert  utils`


<!-- Sent from Reviewable.io -->

jacobdeery reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, let's do it

> Reviewed 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/5#-:-M4Pn9rH8_A2jOXgYdLy:b-twfp5q)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
